ion-c-sys-macros/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use quote::quote;
+use syn::{parse_quote, FnArg, Ident, PatType, ReturnType};
+
+/// This macro is intended to support the `IonCReaderHandle` type. When any
+/// function returns an `IonCError`, the error should be populated with the
+/// current reader position.
+///
+/// Without this macro, errors could be decorated like so:
+///
+/// ```no_test
+/// fn foo() -> IonCResult<()> {
+///     fn1().map_err(|err| include_current_position(self.reader, err))?
+///     // ...
+///     Ok(())
+/// }
+/// ```
+///
+/// However, functions can fail in many places, so decorating each of these with
+/// `map_err` is tedious and error prone:
+///
+/// ```no_test
+/// fn foo() -> IonCResult<()> {
+///   fn1()?.map_err(..);
+///   fn2()?.map_err(..);
+/// }
+/// ```
+///
+/// This macro rewrites the function such that any error gets the position
+/// added. It does this by moving the existing implementation into a closure and
+/// then calling `map_err` on the result.
+///
+/// To make the variety of possible functions work (which include mutable
+/// borrows, lifetime elision), the closure uses the `move` keyword. There are
+/// some gymnastics (adapted from dtolnay/no-panic) to make that work out.
+#[proc_macro_attribute]
+pub fn position_error(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let mut function = syn::parse_macro_input!(item as syn::ItemFn);
+
+    let mut move_self = None;
+    let mut arg_pat = Vec::new();
+    let mut arg_val = Vec::new();
+    for (i, input) in function.sig.inputs.iter_mut().enumerate() {
+        let numbered = Ident::new(&format!("__arg{}", i), Span::call_site());
+        match input {
+            FnArg::Typed(PatType { pat, .. }) => {
+                arg_pat.push(quote!(#pat));
+                arg_val.push(quote!(#numbered));
+                *pat = parse_quote!(mut #numbered);
+            }
+            FnArg::Receiver(_) => {
+                move_self = Some(quote! {
+                    if false {
+                        loop {}
+                        #[allow(unreachable_code)]
+                        {
+                            let __self = self;
+                        }
+                    }
+                });
+            }
+        }
+    }
```

Comment:
The best way to explain this is: the code is munged in arcane ways that dtolnay invented. If you remove one of the arcane steps, things fail to compile. For example, if we remove the `#move_self` in the final codegen, then the methods that return self references like this one:

```
fn get_field_name(&mut self) -> IonCResult<StrSliceRef>;
```

will fail to compile:

```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements
   --> ion-c-sys/src/reader.rs:526:12
    |
526 |         Ok(StrSliceRef::new(self, str_ref))
    |            ^^^^^^^^^^^^^^^^
    |
note: first, the lifetime cannot outlive the lifetime `'_` as defined on the body at 518:5...
   --> ion-c-sys/src/reader.rs:518:5
    |
518 |     #[position_error]
    |     ^^^^^^^^^^^^^^^^^
note: ...so that closure can access `self`
   --> ion-c-sys/src/reader.rs:526:29
    |
526 |         Ok(StrSliceRef::new(self, str_ref))
    |                             ^^^^
note: but, the lifetime must be valid for the anonymous lifetime defined on the method body at 520:20...
   --> ion-c-sys/src/reader.rs:520:20
    |
520 |     fn read_string(&mut self) -> IonCResult<StrSliceRef> {
    |                    ^^^^^^^^^
note: ...so that the expression is assignable
   --> ion-c-sys/src/reader.rs:518:5
    |
518 |     #[position_error]
    |     ^^^^^^^^^^^^^^^^^
    = note: expected `Result<string::StrSliceRef<'_>, _>`
               found `Result<string::StrSliceRef<'_>, _>`
    = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

I don't really know enough about the arcane magic to give it a reasonable comment, hence the sort of handwave one I added already:

```
/// This macro rewrites the function such that any error gets the position
/// added. It does this by moving the existing implementation into a closure and
/// then calling `map_err` on the result.
///
/// To make the variety of possible functions work (which include mutable
/// borrows, lifetime elision), the closure uses the `move` keyword. There are
/// some gymnastics (adapted from dtolnay/no-panic) to make that work out.
```